docs/html/user_guide.rst

```diff
@@ -1306,6 +1306,238 @@ issue tracker`_ if you believe that your problem has exposed a bug in pip.
+Backtracking user guide information
+===================================
+
+Related to `GH issue 8713 <https://github.com/pypa/pip/issues/8713>`__.
+
+Objective
+---------
+
+The purpose of this content is to provide documentation of 1) why this
+occurs, and 2) how the user can possible solve it.
+
+Location
+--------
+
+The end location of this content will be on\ `the pip user
+guide <https://pip.pypa.io/en/latest/user_guide/#id35>`__, somewhere
+near the Resolution Impossible section.
+
+--------------
+
+Dependency resolution backtracking
+==================================
+
+Or more commonly known as *"Why does pip download multiple versions of
+the same package over and over again during an install?"*.
+
+The purpose of this section is to provide explanation of why
+backtracking happens, and practical suggestions to pip users who
+encounter dependency resolution backtracking during a ``pip install``
+command.
+
+What is backtracking?
+---------------------
+
+Backtracking is not a bug, or an unexpected behaviour. It is part of the
+way pip's dependency resolution process works.
+
+During a pip install (e.g. ``pip install tea``), pip needs to work out
+the packages dependencies (e.g. ``spoon``, ``hot-water``, ``cup`` etc),
+and the versions of each of these packages it needs to install.
+
+For each of these dependent packages, it needs to work out which version
+of a package is a good candidate to install.
+
+If everythig goes well, this will result in pip computing a set of
+compatible versions of all these packages.
+
+In the case where a package has a lot of versions, arriving at a good
+candidate can take a lot of time. (The amount of time depends on the
+package size, the number of versions pip must try, among other things)
+
+How does backtracking work?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+pip does this by trying one version, if it finds out it isn’t compatible
+it needs to “go back” (backtrack) and download an older version.
+
+It then tries that version, if it is successful it will continue onto
+the next package - if not it will continue to backtrack until it finds a
+compatible version.
+
+At the end of this process, pip will have the set of compatible
+packages, which it then installs.
+
+This backtrack behaviour can end in 2 ways - either 1) it will
+successfully find a set packages it can install (good news!), or 2) it
+will eventually display `resolution
+impossible <https://pip.pypa.io/en/latest/user_guide/#id35>`__ error
+message (not so good).
+
+If pip starts backtracking during dependency resolution, it does not
+know how long it will backtrack, and how much computation would be
+needed.
+
+Why does backtracking occur?
+----------------------------
+
+With the release of the new resolver (resolver 2020), pip is now more
+strict in the package versions it installs, when a users does a
+``pip install`` command. This new behaviour means that pip works harder
+to find out which version of a package is a good candidate to install.
```

Comment:
This is true.

I also think it would be good here to say: This new behaviour, therefore, reduces the risk that installing a new package will accidentally break an existing installed package so your environment gets messed up.

Also, I think readers will be interested in a deeper reason why pip needs to backtrack: there are millions of packages in the world, both public and private, and there is no all-encompassing index of what packages depend on and require, and the maintainers of packages do not coordinate with each other to make sure things work together. So pip doesn't know whether this new package is going to be compatible with the existing packages/constraints/requirements until it tries. I think this is a reasonable user-readable summary - @pfmoore @pradyunsg @uranusjr ? is it accurate enough?

Reply:
I'd explain that somewhat differently (the "millions of packages" idea seems like a distraction to me).

Pip needs to backtrack because initially, it doesn't have all the information it needs to work out the correct set of packages. This is because package indexes don't provide full package dependency information before you have downloaded the package. So pip needs to start by making assumptions, and check as it goes along. When it finds that an assumption is incorrect, it has to try another approach, which means discarding some of the work that has already been done, and going back to choose another path.

Reply:
> Pip needs to backtrack because initially, it doesn't have all the information it needs to work out the correct set of packages. This is because package indexes don't provide full package dependency information before you have downloaded the package.

I've included this in the *why* section.

> So pip needs to start by making assumptions, and check as it goes along. When it finds that an assumption is incorrect, it has to try another approach, which means discarding some of the work that has already been done, and going back to choose another path.

This reads more as *why* pip needs to do this, so I've moved it there.